ibis/common/selectors.py

```diff
@@ -29,3 +28,79 @@ def expand(self, table: ir.Table) -> Sequence[ir.Value]:
             A sequence of value expressions that match the selector
 
         """
+
+    @abc.abstractmethod
+    def expand_names(self, table: ir.Table) -> frozenset[str]:
+        """Compute the set of column names that match the selector."""
+
+
+class Selector(Expandable):
+    """A column selector."""
+
+    def expand(self, table: ir.Table) -> Sequence[ir.Value]:
+        names = self.expand_names(table)
+        return list(map(table.__getitem__, filter(names.__contains__, table.columns)))
+
+    def __and__(self, other: Selector) -> Selector:
+        """Compute the logical conjunction of two `Selector`s.
+
+        Parameters
+        ----------
+        other
+            Another selector
+        """
+        return And(self, other)
+
+    def __or__(self, other: Selector) -> Selector:
+        """Compute the logical disjunction of two `Selector`s.
```

Comment:
"logical disjunction" is not a phrase I'd expect most users to be familiar with. If we view selectors as generating a set of column names, "union" (here) and "intersection" (for `__and__`) may be a bit clearer.

Reply:
This was pre-existing. It can be changed in a follow-up. I was originally going for precision, and conceptually clarity.